agent/reconnectingpty/buffered.go

```diff
@@ -0,0 +1,238 @@
+				_, err = conn.Write(part)
+				if err != nil {
+					logger.Warn(ctx,
+						"error writing to active connection",
+						slog.F("connection_id", cid),
+						slog.Error(err),
+					)
+					rpty.metrics.WithLabelValues("write").Add(1)
+				}
+			}
+			rpty.mutex.Unlock()
+		}
+	}()
+
+	return rpty
+}
+
+// lifecycle manages the lifecycle of the reconnecting pty.  If the context ends
+// the reconnecting pty will be closed.
+func (rpty *bufferedReconnectingPTY) lifecycle(ctx context.Context, logger slog.Logger) {
+	rpty.timer = time.AfterFunc(attachTimeout, func() {
+		rpty.Close("reconnecting pty timeout")
+	})
+
+	logger.Debug(ctx, "reconnecting pty ready")
+	rpty.state.setState(StateReady, nil)
+
+	state, reasonErr := rpty.state.waitForStateOrContext(ctx, StateClosing)
+	if state < StateClosing {
+		// If we have not closed yet then the context is what unblocked us (which
+		// means the agent is shutting down) so move into the closing phase.
+		rpty.Close(reasonErr.Error())
+	}
+	rpty.timer.Stop()
+
+	rpty.mutex.Lock()
+	defer rpty.mutex.Unlock()
+
+	rpty.circularBuffer.Reset()
+
+	// Log these closes only for debugging since the connections or processes
+	// might have already closed on their own.
+	for _, conn := range rpty.activeConns {
+		err := conn.Close()
+		if err != nil {
+			logger.Debug(ctx, "closed conn with error", slog.Error(err))
+		}
+	}
+
+	err := rpty.ptty.Close()
+	if err != nil {
+		logger.Debug(ctx, "closed ptty with error", slog.Error(err))
+	}
+
+	err = rpty.process.Kill()
+	if err != nil {
+		logger.Debug(ctx, "killed process with error", slog.Error(err))
+	}
+
+	logger.Debug(ctx, "closed reconnecting pty")
+	rpty.state.setState(StateDone, xerrors.Errorf("reconnecting pty closed: %w", reasonErr))
+}
+
+func (rpty *bufferedReconnectingPTY) Attach(ctx context.Context, connID string, conn net.Conn, height, width uint16, logger slog.Logger) error {
+	// This will kill the heartbeat once we hit EOF or an error.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	logger.Debug(ctx, "reconnecting pty attach")
+	err := rpty.doAttach(ctx, connID, conn, height, width, logger)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		rpty.mutex.Lock()
+		defer rpty.mutex.Unlock()
+		delete(rpty.activeConns, connID)
+	}()
+
+	// Pipe conn -> pty and block.  pty -> conn is handled in newBuffered().
+	readConnLoop(ctx, conn, rpty.ptty, rpty.metrics, logger)
+	return nil
+}
+
+// doAttach adds the connection to the map, replays the buffer, and starts the
+// heartbeat.  It exists separately only so we can defer the mutex unlock which
+// is not possible in Attach since it blocks.
+func (rpty *bufferedReconnectingPTY) doAttach(ctx context.Context, connID string, conn net.Conn, height, width uint16, logger slog.Logger) error {
+	// Ensure we do not write to or close connections while we attach.
+	rpty.mutex.Lock()
+	defer rpty.mutex.Unlock()
+
+	state, err := rpty.state.waitForStateOrContext(ctx, StateReady)
```

Comment:
This makes sense and I had the same concerns.  My thinking though is that I would need to hold the mutex while checking the state otherwise if the pty closes in the meantime I will end up adding it to the map and the conn might not get closed.

Assuming my thinking is correct, maybe this means I should refactor to avoid using a map in the first place.

Reply:
Actually I think I can fix this by adding a new goroutine that closes the conn when the state changes to closed.

Reply:
That's a good point.  If you keep the map, it indicates to me that there should only be one lock, not two.  Since you should only insert into the map if the reconnectingPTY is in a specific state, both the map and the state should be covered by a single lock.  That also dodges issues of deadlocking, and since waiting for a `Cond` releases the lock while you are waiting, it also avoids the performance issues.

Reply:
Great idea!  I implemented this with a function passed to `waitForStateOrContext()` so I could manipulate the map in that function while holding the `Cond` lock.  It feels a bit JavaScript-y to me and I am not sure if it looks like idiomatic Go so please let me know if you think it looks weird.